README.md
# wiringpi-test

It is a package that helps test pins with [wiringPi](https://github.com/wiringPi).<br>
<br>

If you want to test gpio input/output functions,<br>
Short all 40 pins (excluding voltage and GND) exposed in Target board.<br>
A pin with a non-gpio function selected as mux is skipped.<br>

![wiringpi-test circuit](wiringpi-test.png)

### Run
![wiringpi-test](wiringpi-test.gif)

### Requirements & Support

I have completed the basic gpio function.<br>
i2c and spi are thinking about generic testing.<br>
Because they are depending on the device they connect to.

```
<Support Board>
- odroid
- raspberry pi

<Support Functions>
- gpio
- i2c
- spi

<Requirements>
- libwiringpi2
- libwiringpi-dev
- wiringpi
```

### Usage in Ubuntu/Debian

Ubuntu/Devian Images support as a package.<br>
raspberry OS is not yet supported.<br>
<br>
Install
```
$ sudo add-apt-repository ppa:how2flow/ppa
$ sudo apt update
$ sudo apt install wiringpi-test
```

Run
```
$ wiringpi-test [ --gpio | --help | --i2c | --spi ]
```

### Usage in Rasbperry OS

This method is not only for Rasbperry OS, which is not yet supported as a package,<br>
but also for testing the wiringpi that has been personally changed.<br>
<br>
Install
```
$ git clone https://github.com/how2flow/wiringpi-test
$ cd wiringpi-test/test
$ make
```

Run
```
$ ./test [gpio | i2c | spi]
```

### Issues

```
blink screen when print gpio pins
- It is disappeared when used with ssh.
```

#### contact

e-mail: steve@how2flow.net


debian/changelog
wiringpi-test (1.0.11) stable; urgency=medium

  * Update README (+ wiringpi-test.png)

 -- Steve Jeong <steve@how2flow.net>  Fri, 08 Sep 2023 06:55:33 +0000

wiringpi-test (1.0.10) stable; urgency=medium

  * Update README (+ wiringpi-test.gif)

 -- Steve Jeong <steve@how2flow.net>  Mon, 21 Aug 2023 01:34:18 +0000

wiringpi-test (1.0.9) stable; urgency=medium

  * Update README 

 -- Steve Jeong <steve@how2flow.net>  Fri, 18 Aug 2023 08:04:03 +0000

wiringpi-test (1.0.8) stable; urgency=medium

  * feat: Add test for Raspberry pi and wiringPi under development

 -- Steve Jeong <steve@how2flow.net>  Fri, 18 Aug 2023 01:54:21 +0000

wiringpi-test (1.0.7) stable; urgency=medium

  * feat: Update help message

 -- Steve Jeong <steve@how2flow.net>  Fri, 11 Aug 2023 04:02:14 +0000

wiringpi-test (1.0.6) stable; urgency=medium

  * fix: screen blink

 -- Steve Jeong <steve@how2flow.net>  Fri, 11 Aug 2023 03:43:08 +0000

wiringpi-test (1.0.5) stable; urgency=medium

  * feat: Support alt functions: i2c & spi 

 -- Steve Jeong <steve@how2flow.net>  Fri, 11 Aug 2023 03:39:03 +0000

wiringpi-test (1.0.4) stable; urgency=medium

  * feat: Remove default option
  * feat: scripts/functions: Remove unnecessary script functions

 -- Steve Jeong <steve@how2flow.net>  Wed, 26 Jul 2023 09:25:37 +0000

wiringpi-test (1.0.3) stable; urgency=medium

  * fix: build error in launchpad
  * (multiple definition)

 -- Steve Jeong <steve@how2flow.net>  Wed, 26 Jul 2023 05:58:32 +0000

wiringpi-test (1.0.2) stable; urgency=medium

  * fix: build error in launchpad

 -- Steve Jeong <steve@how2flow.net>  Wed, 26 Jul 2023 05:42:36 +0000

wiringpi-test (1.0.1) stable; urgency=medium

  * debian: Add postrm script

 -- Steve Jeong <steve@how2flow.net>  Wed, 26 Jul 2023 04:54:40 +0000

wiringpi-test (1.0.0) stable; urgency=medium

  * 1.0.0 alpha version.
  * support gpio only

 -- Steve Jeong <steve@how2flow.net>  Wed, 26 Jul 2023 03:53:00 +0000

wiringpi-test (0.0.1) unstable; urgency=medium

  * initial release.

 -- Steve Jeong <steve@how2flow.net>  Tue, 25 Jul 2023 11:15:14 +0900


src/i2c.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>

#include "wiringpitest.h"

static int setup;
static int fd;

static int setup_i2c(int address, char *bus)
{
	char *device = (char *)malloc(sizeof(char) * 11);

	if (wiringPiSetup() == -1)
		exit(1);

	snprintf(device, 11, "%s%s", "/dev/i2c-", bus);
	fd = wiringPiI2CSetupInterface(device, address);

	return 1;
}

int store_i2c(void *args)
{
	struct i2c_args *arguments = (struct i2c_args *) args;
	/* int data; */

	if (!setup) {
		setup = setup_i2c(arguments->addr, arguments->bus);
	}

	// TODO
	// Is there a general test routine?
	/*
	srand(time(NULL));
	data = rand() % 2;

	wiringPiI2CWrite(fd, data);
	*/

	delay(500);

	return fd;
}


src/spi.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wiringPi.h>
#include <wiringPiSPI.h>

#include "wiringpitest.h"

#define MAX_SIZE (1024*1024)

static int setup;
static int fd;

static int setup_spi(int cs, int speed)
{
	if (wiringPiSetup() == -1)
		exit(1);

	fd = wiringPiSPISetup(cs, speed);

	return 1;
}

int store_spi(void *args)
{
	struct spi_args *arguments = (struct spi_args *) args;
	unsigned char *data = (unsigned char *)malloc(MAX_SIZE);

	if (!setup) {
		setup = setup_spi(arguments->cs, arguments->speed);
	}

	if (wiringPiSPIDataRW(arguments->cs, data, MAX_SIZE) == -1) {
		perror("Failed read/write spidev");
		return -1;
	}

	return fd;
}


src/main.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "wiringpitest.h" 

static struct i2c_args *i2c_info;
static struct spi_args *spi_info;

static int fd;

static void *bg_gpio_thread(void *param)
{
	int ret;

	ret = store_gpio();
	if (ret < 0) {
		printf("Failed using gpio\n");
	}

	return NULL;
}

static void *bg_i2c_thread(void *param)
{
	int ret;

	fd = store_i2c(param);
	if ((ret = fd) < 0) {
		printf("Failed using i2c\n");
	}

	return NULL;
}

static void *bg_spi_thread(void *param)
{
	int ret;

	fd = store_spi(param);
	if ((ret = fd) < 0) {
		printf("Failed using spi\n");
	}

	return NULL;
}

int main(int argc, char *argv[])
{
	pthread_t gpio, i2c, spi;
	int gpioTh, i2cTh, spiTh;
	int param, ret;
	char *cmd = (char *)malloc(sizeof(char) * 15);

	if (argc < 2) {
		printf("Need param Call \"wiringpi-test --help\"\n");
		return -1;
	}

	if(!strcmp(argv[1], "gpio")) {
		param = 100;

		gpioTh = pthread_create(&gpio, NULL, bg_gpio_thread, &param);
		if (gpioTh < 0) {
			perror("Failed create gpio thread");
			exit(0);
		}

		show_header();
	}

	if(!strcmp(argv[1], "i2c")) {
		snprintf(cmd, 15, "%s%s", "i2cdetect -y ", argv[2]);
		ret = system(cmd);
		if (ret < 0) {
			perror("system cmd error");
		}

		printf("Input i2c device address: ");
		ret = scanf("%x", &param);
		if (ret < 0) {
			perror("input error");
		}

		i2c_info->addr = param;
		i2c_info->bus = argv[2];

		i2cTh = pthread_create(&i2c, NULL, bg_i2c_thread, (void *) i2c_info);
		if (i2cTh < 0) {
			perror("Failed create i2c thread");
			exit(0);
		}

		show_i2cbus(fd);
	}

	if(!strcmp(argv[1], "spi")) {
		printf("Input spi speed (default: 1000000): ");
		ret = scanf("%d", &param);
		if (ret < 0) {
			perror("input error");
		}

		if (param < SPI_SPEED) {
			spi_info->speed = SPI_SPEED;
		}
		else {
			spi_info->speed = param;
		}

		spi_info->cs = atoi(argv[2]);

		spiTh = pthread_create(&spi, NULL, bg_spi_thread, (void *) spi_info);
		if (spiTh < 0) {
			perror("Failed create spi thread");
			exit(0);
		}

		show_spidata();
	}

	close(fd);

	return 0;
}


src/show.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>

#include "wiringpitest.h" 

int show_header(void)
{
	float cnt = 0;
	int ret;

	while (cnt < 60) {
		ret = system("clear");
		if (ret < 0) {
			perror("failed");
		}

		ret = system("gpio readall");
		if (ret < 0) {
			perror("failed");
		}

		cnt += 0.1;
		printf("time: %.1f\n", cnt);
		usleep(100000);
	}

	return 0;
}

int show_i2cbus(int fd)
{
	float cnt = 0;
	int data, ret;

	while (cnt < 60) {
		ret = system("clear");
		if (ret < 0) {
			perror("failed");
		}

		data = wiringPiI2CRead(fd);
		printf("Read: 0x%x\n", data);

		cnt += 0.1;
		printf("time: %.1f\n", cnt);
		usleep(100000);
	}

	return 0;
}

int show_spidata(void)
{
	float cnt = 0;
	int ret;

	while(cnt < 60) {
		ret = system("clear");
		if (ret < 0) {
			perror("failed");
		}

		ret = system("spi-pipe -d /dev/spidev* -b 4 -n 40 < /dev/zero | hexdump -C");
		if (ret < 0) {
			perror("failed");
		}

		cnt += 0.1;
		printf("time: %.1f\n", cnt);
		usleep(100000);
	}

	return 0;
}


src/pinmap.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
	while (1) {
		system("clear && sync");
		system("gpio readall");

		usleep(100000);
	}

	return 0;
}


include/wiringpitest.h
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#ifndef __WIRINGPITEST_H__
#define __WIRINGPITEST_H__

/* vars */

/* functions */
int show_header(void);

int store_gpio(void);
#endif


src/gpio.c
/*******************************************************************************
 Copyright (C) 2022-2023 Steve Jeong <steve@how2flow.net>
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <wiringPi.h>

#include "wiringpitest.h"

static int setup;

/* mapping */
static int phyToPin[64] = {
	-1,
	-1, -1,
	 8, -1,
	 9, -1,
	 7, 15,
	-1, 16,
	 0,  1,
	 2, -1,
	 3,  4,
	-1,  5,
	12, -1,
	13,  6,
	14, 10,
	-1, 11,
	30, 31,
	21, -1,
	22, 26,
	23, -1,
	24, 27,
	-1, -1,
	-1, -1,
	/* padding */
	-1, -1, -1, -1, -1, -1, -1, -1,
	-1, -1, -1, -1, -1, -1, -1, -1,
	-1, -1, -1, -1, -1, -1, -1
};

static int setup_gpio(void)
{
	int phy; // physical header pin number
	int pin; // wiringpi pin number

	wiringPiSetup();

	for (phy = 1; phy <= 40; phy++) {
		pin = phyToPin[phy];

		if (pin == -1)
			continue;

		if (getAlt(pin) > 1)
			continue;

		pinMode(pin, INPUT);
	}

	return 1;
}

static void toggle_gpio(int pin)
{
		digitalWrite(pin, HIGH);
		delay(500);
		digitalWrite(pin, LOW);
		delay(500);
}

static void use_gpio(void)
{
	int phy;
	int pin;

	for (phy = 1; phy <= 40; phy++) {
		pin = phyToPin[phy];

		if (pin == -1)
			continue;

		if (getAlt(pin) > 1)
			continue;

		pinMode(pin, OUTPUT);
		toggle_gpio(pin);
		pinMode(pin, INPUT);
	}
}

int store_gpio(void)
{
	int phy; // physical header pin number
	int pin; // wiringpi pin number

	if (!setup) {
		setup = setup_gpio();
	}

	for (phy = 1; phy <= 40; phy++) {
		pin = phyToPin[phy];

		if (pin == -1)
			continue;

		if (getAlt(pin) > 1)
			continue;

		use_gpio();
	}

	return 0;
}
